File: core/java/src/main/java/hara/lang/protocol/IStreamWrite.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Asynchronous writable stream capability. */
@HaraProtocolBinding(
    namespace = "std.protocol.istreamwrite",
    name = "IStreamWrite",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IStreamWrite {
  @HaraMethod(value = "write", arity = 2)
  Object write(Object value);
}

File: core/rust/src/lang/protocol/istreamwrite.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istreamwrite",
    name = "IStreamWrite",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IStreamWrite {
    type Value;
    type Output;

    #[hara_method(value = "write", arity = 2)]
    fn write(&mut self, value: Self::Value) -> Self::Output;
}
